### ISSUE TITLE
`header.documents(fill=True)` is caching data?

### ISSUE THREAD

**Issue by CJ-Wright** [bluesky/databroker#342]
My understanding of the documents being held in the generator was to prevent the data from taking over all the memory. However, when I load up 54 images via a for loop it seems that my memory piles up, to the tune of 7GB just looping through the documents. Is this operating properly?
```python
db_path = '/media/christopher/DATA/Research/Columbia/pal/db'
config = {'description': 'lightweight personal database',
          'metadatastore': {'module': 'databroker.headersource.sqlite',
                            'class': 'MDS',
                            'config': {'directory': db_path,
                                       'timezone': 'US/Eastern'}},
          'assets': {'module': 'databroker.assets.sqlite',
                     'class': 'Registry',
                     'config': {'dbpath': db_path + '/database.sql'}}}
for n, d in db[-1].documents(fill=True):
    # source.emit((n, d))
    if n == 'event':
        print(d['data']['image])
```
I think (I used real data to produce this) this databroker can be created by
```python
from databroker.broker import Broker
import uuid
import numpy as np
import os
from bluesky.utils import new_uid
import time

class NpyWriter:
    """
    Each call to the ``write`` method saves a file and creates a new filestore
    resource and datum record.
    """

    SPEC = 'npy'

    def __init__(self, fs, root):
        self._root = root
        self._closed = False
        self._fs = fs
        # Open and stash a file handle (e.g., h5py.File) if applicable.

    def write(self, data):
        """
        Save data to file, generate and insert new resource and datum.
        """
        if self._closed:
            raise RuntimeError('This writer has been closed.')
        fp = '{}.npy'.format(str(uuid.uuid4()))
        np.save(os.path.join(self._root, fp), data)
        resource = self._fs.insert_resource(self.SPEC, fp, resource_kwargs={},
                                            root=self._root)
        datum_id = str(uuid.uuid4())
        self._fs.insert_datum(resource=resource, datum_id=datum_id,
                              datum_kwargs={})
        return datum_id

    def close(self):
        self._closed = True

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


db_path = '/media/christopher/DATA/Research/Columbia/pal/db'
config = {'description': 'lightweight personal database',
          'metadatastore': {'module': 'databroker.headersource.sqlite',
                            'class': 'MDS',
                            'config': {'directory': db_path,
                                       'timezone': 'US/Eastern'}},
          'assets': {'module': 'databroker.assets.sqlite',
                     'class': 'Registry',
                     'config': {'dbpath': db_path + '/database.sql'}}}

db = Broker.from_config(config)
writer = NpyWriter(db.fs, db_path)

def run():
    start = {}
    descriptor = {}
    for i in range(55):
        if not start:
            suid = new_uid()
            start = {'uid': suid, 'time': time.time()}
            yield 'start', start
        if not descriptor:
            duid = new_uid()
            descriptor = {'uid': duid,
                          'name': 'primary',
                          'run_start': suid,
                          'data_keys': {
                              'image': {
                                  'source': 'sidewinder', 'dtype': 'array',
                                  'shape': [2880, 2880], 'unit': 'NA'}},
                          'time': time.time()}
            yield 'descriptor', descriptor
        yield 'event', {'uid': new_uid(),
                         'descriptor': duid,
                         'filled': {'image': True},
                         'data': {'image': np.random.random((2880, 2880))},
                         'timestamps': {'image': time.time(),},
                         'seq_num': i,
                         'time': time.time(),
                         }
    yield 'stop', {'uid': new_uid(), 'run_start': suid,
                   'time': time.time()}

for n, d in run():
    if n == 'descriptor':
        d['data_keys']['image']['external'] = True
    if n == 'event':
        d['data']['image'] = writer.write(d['data']['image'])
        d['filled']['image'] = False
    db.insert(n, d)
```

**Comment by jrmlhermitte** [comment]
could this be related to what happened in #299? @danielballan

**Comment by tacaswell** [comment]
Check the handlers are not caching the data.

**Comment by CJ-Wright** [comment]
This is using the [`NpyHandler`](https://github.com/NSLS-II/databroker/blob/eace7fc9c77a2dcef8396aad82ccc1ccc30d655a/databroker/assets/handlers.py#L407) which looks clean to me, although as second set of eyes would be helpful.

**Comment by danielballan** [comment]
I'm on it.

**Comment by danielballan** [comment]
Generated the data and verified that it seems to all be there:

```
$ ipython -i /tmp/cj2.py
```

```python
In [1]: len(list(db()))
Out[1]: 1

In [7]: Counter(1 for _ in db[-1].events())
Out[7]: Counter({1: 55})
```

Ran your script that _should_ stream the data through without accumulating refs in memory

```
In [12]: %run -i /tmp/cj1.py
<printed a lot of arrays>
```

Checked [memory usage](https://stackoverflow.com/questions/938733/total-memory-used-by-python-process):

```python
In [16]: import os
    ...: import psutil
    ...: process = psutil.Process(os.getpid())
    ...: print(process.memory_info().rss)
    ...:
16130048
```

That's not so high, nowhere near 7 GB.  But, granted, this check was performed _after_ the printing was all done. Is the garbage collector was having trouble keeping up _during_ the loop? I checked that too:

```python
usage_monitor = []
for n, d in db[-1].documents(fill=True):
    # source.emit((n, d))
    if n == 'event':
       print(d['data']['image'])
       usage_monitor.append(process.memory_info().rss)
```

The memory usage never exceeded 106 MB.

```
In [19]: max(usage_monitor)
Out[19]: 106106880
```

So it looks like I can't reproduce your problem. Can you add these usage checks on your machine and share the results?

**Comment by danielballan** [comment]
P.S. Thanks for this really good SSCCE! All I had to change was `db_path = '/media/christopher/DATA/Research/Columbia/pal/db'` to `db_path = /tmp/cj` to get it running on my machine. It made it very easy to try to help.

**Comment by CJ-Wright** [comment]
Edit: I take that back (I was running with`fill=False`). 
Here is what i ran to make the databroker:
```python
from databroker.broker import Broker
import uuid
import numpy as np
import os
from bluesky.utils import new_uid
import time

class NpyWriter:
    """
    Each call to the ``write`` method saves a file and creates a new filestore
    resource and datum record.
    """

    SPEC = 'npy'

    def __init__(self, fs, root):
        self._root = root
        self._closed = False
        self._fs = fs
        # Open and stash a file handle (e.g., h5py.File) if applicable.

    def write(self, data):
        """
        Save data to file, generate and insert new resource and datum.
        """
        if self._closed:
            raise RuntimeError('This writer has been closed.')
        fp = '{}.npy'.format(str(uuid.uuid4()))
        np.save(os.path.join(self._root, fp), data)
        resource = self._fs.insert_resource(self.SPEC, fp, resource_kwargs={},
                                            root=self._root)
        datum_id = str(uuid.uuid4())
        self._fs.insert_datum(resource=resource, datum_id=datum_id,
                              datum_kwargs={})
        return datum_id

    def close(self):
        self._closed = True

    def __enter__(self):
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.close()


db_path = '/media/christopher/DATA/Research/Columbia/temp_db'
config = {'description': 'lightweight personal database',
          'metadatastore': {'module': 'databroker.headersource.sqlite',
                            'class': 'MDS',
                            'config': {'directory': db_path,
                                       'timezone': 'US/Eastern'}},
          'assets': {'module': 'databroker.assets.sqlite',
                     'class': 'Registry',
                     'config': {'dbpath': db_path + '/database.sql'}}}

db = Broker.from_config(config)
writer = NpyWriter(db.fs, db_path)

def run():
    start = {}
    descriptor = {}
    for i in range(10):
        if not start:
            suid = new_uid()
            start = {'uid': suid, 'time': time.time()}
            yield 'start', start
        if not descriptor:
            duid = new_uid()
            descriptor = {'uid': duid,
                          'name': 'primary',
                          'run_start': suid,
                          'data_keys': {
                              'image': {
                                  'source': 'sidewinder', 'dtype': 'array',
                                  'shape': [2880, 2880], 'unit': 'NA'}},
                          'time': time.time()}
            yield 'descriptor', descriptor
        yield 'event', {'uid': new_uid(),
                         'descriptor': duid,
                         'filled': {'image': True},
                         'data': {'image': np.random.random((2880, 2880)).astype('uint16')},
                         'timestamps': {'image': time.time(),},
                         'seq_num': i,
                         'time': time.time(),
                         }
    yield 'stop', {'uid': new_uid(), 'run_start': suid,
                   'time': time.time()}

for n, d in run():
    if n == 'descriptor':
        d['data_keys']['image']['external'] = True
    if n == 'event':
        d['data']['image'] = writer.write(d['data']['image'])
        d['filled']['image'] = False
        print(d['seq_num'])
    db.insert(n, d)
```
(Note only ten images)

Here is what I ran to check the data:
```python
import copy

from databroker import Broker

db_path = '/media/christopher/DATA/Research/Columbia/temp_db'
config = {'description': 'lightweight personal database',
          'metadatastore': {'module': 'databroker.headersource.sqlite',
                            'class': 'MDS',
                            'config': {'directory': db_path,
                                       'timezone': 'US/Eastern'}},
          'assets': {'module': 'databroker.assets.sqlite',
                     'class': 'Registry',
                     'config': {'dbpath': db_path + '/database.sql'}}}

db = Broker.from_config(config)
# db.prepare_hook = lambda name, doc: copy.deepcopy(doc)

usage_monitor = []
import os
import psutil
process = psutil.Process(os.getpid())

for n, d in db[-1].documents(
        fill=True
):
    if n == 'event':
        print('{} MB'.format(str(process.memory_info().rss/1e6)))
        usage_monitor.append(process.memory_info().rss)
```
and the output
```sh
160.33792 MB
177.016832 MB
193.744896 MB
210.444288 MB
227.119104 MB
244.088832 MB
260.816896 MB
277.487616 MB
294.191104 MB
310.890496 MB
```
Note that while the memory outlay is small it is increasing.
I'm running a databroker installed from source at commit 3c8ce3fee0d8981786ebe94ca27fe2b15fbf0117

~~Interesting, so I ran my example and got the same results as you, but then I loaded the databroker I made with the hdf5 data and got the cache (the same inspection of the databroker for both).~~

**Comment by danielballan** [comment]
> but then I loaded the databroker I made with the hdf5 data and got the cache (the same inspection of the databroker for both).

"Got the cache"?

**Comment by CJ-Wright** [comment]
Sorry got memory increasing

**Comment by danielballan** [comment]
When I plot `usage_monitor`, I do not see memory increasing:

![process-mem-usage](https://user-images.githubusercontent.com/2279598/35991974-e7cb72d0-0cd6-11e8-8cf9-954759012fef.png)

This looks like what I would expect: the somewhat opaque action of Python garbage collection is generally keeping the memory usage low, with occasional jumps. I'm not convinced there is a problem in databroker (yet). If you are worried about runaway memory usage, you could try inserting `gc.collect()` inside your loop. Does that help?

**Comment by CJ-Wright** [comment]
I added the collector
```python
...
import gc
process = psutil.Process(os.getpid())

for n, d in db[-1].documents(fill=True):
    if n == 'event':
        print('{} MB'.format(str(process.memory_info().rss/1e6)))
        usage_monitor.append(process.memory_info().rss)
        gc.collect()
import matplotlib.pyplot as plt
plt.plot(usage_monitor)
plt.show()
```
Same result
![figure_1-21](https://user-images.githubusercontent.com/6740689/35992371-1ce6353a-0cd8-11e8-899a-95ec52ddb0e1.png)

**Comment by danielballan** [comment]
What could be different about what you are doing and what I am going? Relevant info:

* I am using the same commit of databroker that you are.
* I'm running the code on my MacBook.

I have not closely read the second lengthy example that you posted; I'm still working from the first example (with `usage_monitor` added and the `db_path` aimed at tmp). Have you changed something important about what you are doing? You mentioned something about "HDF5 data" that I didn't follow, and I'm assuming that's not important. Let's focus on that first example you posted.

**Comment by CJ-Wright** [comment]
Yea, sorry the HDF5 stuff doesn't matter.

**Comment by CJ-Wright** [comment]
Same result in `/tmp/temp_db`
160.690176 MB
177.434624 MB
194.162688 MB
211.103744 MB
227.807232 MB
244.506624 MB
261.206016 MB
277.909504 MB
294.608896 MB
311.308288 MB

**Comment by danielballan** [comment]
To ensure that we are doing exactly the same thing, I have consolidated your example into one Python script that generates a fresh (randomly named) database each time -- and cleans up afterward -- and generates a PNG of the memory usage plot.

https://gist.github.com/danielballan/2a474eee53a52fcd8716d2fa451bd4a5

Example result on my system with this Python:

```
$ python --version
Python 3.6.4 :: Anaconda, Inc.
```

![mem-usage-tmpk3v2avhf](https://user-images.githubusercontent.com/2279598/35994195-b407e85a-0cdd-11e8-8340-d140749f9054.png)

**Comment by CJ-Wright** [comment]
I ran your script (with two modifications (I changed the number of images to 10 and the dtype to 'uint16') because I ran out of memory when writing to temp).
Here is my graph
![mem-usage-tmp6bn4xq7f](https://user-images.githubusercontent.com/6740689/35994674-2ca6cdb6-0cdf-11e8-8fbb-f36eba7118c1.png)

This is weird I don't know if this breaks anything out at the ring. It would be interesting to run this on a NSLS-II computer.

**Comment by danielballan** [comment]
Which version of Python? Also, I suggest decreasing the array size and increasing the number of images, just to be sure you are giving the garbage collector time to catch on.

**Comment by CJ-Wright** [comment]
py3.5 I will try to reduce the size and run again.

**Comment by danielballan** [comment]
It is probably also worth checking whether gc was improved between 3.5 and 3.6.

**Comment by tacaswell** [comment]
you can try throwing some explicit `gc.collect()` calls in too?

My understanding is that things should be collected immediately when their reference counts hit 0.

**Comment by CJ-Wright** [comment]
Similar results with the gc. (note there are 55 images at default numpy dtype at size (200, 200))
![mem-usage-tmpjsgd9tnj](https://user-images.githubusercontent.com/6740689/35998977-4f37cfda-0cec-11e8-913e-3767d40e425d.png)

Is there a good way to ask an object how many references it has/where those references are?
Edit: I'll use the  gc reffers to see if something is going on https://docs.python.org/3/library/gc.html#gc.get_referrers

**Comment by danielballan** [comment]
http://bfy.tw/GVHX 😈

**Comment by danielballan** [comment]
Ah, I see you got there yourself and edited your last comment. Good work.

**Comment by CJ-Wright** [comment]
@tacaswell and @jrmlhermitte can you try to reproduce? Tim and I have been able to reproduce this. (please try to run with both the normal sized and smaller images since I can't hold all the images at once)

**Comment by danielballan** [comment]
Have you tried in a Python 3.6 env?

**Comment by CJ-Wright** [comment]
@chiahaoliu  ran in py36

**Comment by jrmlhermitte** [comment]
same for me using your gist.
Only modification i made is:
```py
from bluesky.tests.utils import temp_config
db = Broker.from_config(temp_config())
```

![memusage](https://user-images.githubusercontent.com/705366/36001312-ec270b4c-0cf3-11e8-9bb5-ccb409a47523.png)



```
(py3) lhermitte@lhermitte-XPS-13-9343:~$ python --version
Python 3.6.2 :: Continuum Analytics, Inc.

In [7]: databroker.__version__
Out[7]: '0.9.4'


```

**Comment by tacaswell** [comment]
![so](https://user-images.githubusercontent.com/199813/36010207-4583fb9e-0d1e-11e8-85ea-fd8e65b83b44.png)


```
In [15]: databroker.__version__
Out[15]: '0.9.4+20.g9d3163f0.dirty'


In [16]: import sys

In [17]: sys.version
Out[17]: '3.7.0a4+ (heads/master:3a0cf93179, Jan 17 2018, 10:26:52) \n[GCC 7.2.1 20171224]'

```

**Comment by tacaswell** [comment]
The issue is that at the very very bottom the sqlite `find` https://github.com/NSLS-II/databroker/blob/3c8ce3fee0d8981786ebe94ca27fe2b15fbf0117/databroker/headersource/sqlite.py#L118-L144

is not lazy.

The real question is why this was working for @danielballan !

Applying the patch from https://github.com/NSLS-II/databroker/pull/344 gives

![so](https://user-images.githubusercontent.com/199813/36011292-2bdf0b74-0d24-11e8-9c32-8fca17150533.png)

**Comment by danielballan** [comment]
Perhaps memory usage does not include swap usage and I was effectively plotting the OS moving data to swap space.

**Comment by jrmlhermitte** [comment]
wow, @tacaswell nice catch!!!!

@danielballan not surprising that Apple hides their information :-P

**Comment by danielballan** [comment]
Closed by #344.
